File: unique-paths-ii.cc
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int> > &obstacleGrid) {
        if (obstacleGrid.empty() || obstacleGrid[0].empty()) return 0;
        int n = obstacleGrid.size();
        int m = obstacleGrid[0].size();
        vector<vector<int> > num(n, vector<int>(m));
        for (int i = 0; i < n && obstacleGrid[i][0] == 0; i++)
            num[i][0] = 1;
        for (int i = 0; i < m && obstacleGrid[0][i] == 0; i++)
            num[0][i] = 1;
        for (int i = 1; i < n; i++)
            for (int j = 1; j < m; j++)
                if (obstacleGrid[i][j] == 0)
                    num[i][j] = num[i - 1][j] + num[i][j - 1];
        return num[n - 1][m - 1];
    }
};

File: unique-paths-ii.java
public class Solution {
    public int uniquePathsWithObstacles(int[][] f) {
        int rows = f.length;
        int cols = f[0].length;
        f[0][0] = 1 - f[0][0];
        for (int i = 1; i < cols; i++)
            f[0][i] = f[0][i - 1] == 0 ? 0 : 1 - f[0][i];
        for (int i = 1; i < rows; i++)
            f[i][0] = f[i - 1][0] == 0 ? 0 : 1 - f[i][0];
        for (int i = 1; i < rows; i++)
            for (int j = 1; j < cols; j++)
                f[i][j] = f[i][j] == 1 ? 0 : f[i - 1][j] + f[i][j - 1];
        return f[rows - 1][cols - 1];
    }
}
